extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/internal/TradfriCoapClient.java

```diff
@@ -22,10 +26,38 @@
 public class TradfriCoapClient extends CoapClient {
 
     private static final int TIMEOUT = 2000;
+    private static final int DEFAULT_DELAY_MILLIS = 500;
+    private static final String STOP_EXECUTOR = "__STOP_EXECUTOR__";
+
+    private final Logger logger;
+    private final LinkedBlockingQueue<PayloadCallbackPair> commandsQueue;
+    private final Thread commandExecutorThread;
 
     public TradfriCoapClient(URI uri) {
         super(uri);
         setTimeout(TIMEOUT);
+        logger = LoggerFactory.getLogger(getClass());
+
+        commandsQueue = new LinkedBlockingQueue<>();
+        commandExecutorThread = new Thread(() -> {
```

Comment:
Why do you create a new Thread on your own? According to our [guidelines](https://www.eclipse.org/smarthome/documentation/development/guidelines.html#d-runtime-behavior) this is not allowed. Please use the `scheduler` and its `submit` method to have it run immediately.

Also if this is now the second binding after the Lifx one that implements such a throttle mechanism, whether it might be better to offer a general throttling implementation that can be used by both bindings?

Reply:
> Why do you create a new Thread on your own?

Using scheduler is suitable for one time jobs or short actions (in a way fire and forget). The idea of this thread is to be a long running operation (start it in the initialization of the handler and stops it on dispose). 

The scheduler will also introduce some limitations.
1. you can't check the state (if it's running or not)
2. You can't restart it (unless you submit a new job - which will have new instance of the command queue)
3. You can't wait for it to finish

Reply:
Keeping a thread occupied with nothing for most of the time is not a good idea though. Especially not on constrained hardware. 

How about something like this:

```java
private final Queue<Command> queue = //...
private Future<?> job;

void sendCommand(Command cmd) {
    queue.add(cmd);
    if (job == null || job.isDone()) {
        job = scheduler.submit(() -> {
            while (!queue.isEmpty()) {
                Command command = queue.poll();
                
                // do send the command...
                
                try {
                    Thread.sleep(200);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        });
    }
}

void destroy() {
    if (job != null) {
        job.cancel(true);
    }
}
```

Reply:
Shouldn't we use a blocking queue instead of that thread.sleep usage?

Reply:
> How about something like this

I guess it is also possible but the if statement: `job == null || job.isDone()` is not atomic which means that in some rare cases we might skip running the job and the command will not be sent.

>Shouldn't we use a blocking queue instead of that thread.sleep usage?

The thread.sleep has nothing to do with the blocking, it's used to allow the device to process the command before sending a new one.

Reply:
> Shouldn't we use a blocking queue

If you can make the queue block for _at least_ 200ms...? I don't think this works though. But maybe I didn't fully get your idea...

> but the if statement: job == null || job.isDone() is not atomic

Of course not, you are right! The above code was meant to illustrate the mechnism ("something _like_ this"), not the final implementation. Otherwise I would have created a PR myself, right? Please add proper synchronization, logging and whatever else you see is necessary in order to make this robust and stable.

Reply:
@wborn since you mentioned that the Lifx uses a throttling mechanism and @kaikreuzer 
 mentioned that @kgoderis had to use such a mechanism for the Tesla binding because the web API locks you out if you fire too many requests at once, do you guys think that we should come up with a general solution and provide it in the framework? Do you think it would be possible to generalize all use cases that we currently have?

Reply:
Yes I've also been thinking about having a general solution in the framework. 

The LIFX lights can be addressed via unicast and multicast (UDP broadcast). The LIFX throttling implementation makes sure that a lock is obtained on all lights or single lights. What I don't like about the implementation is that it uses Thread.sleep. But since the packet interval is just 50ms it this is not really an issue. 

The Nest API also has [rate limits](https://developers.nest.com/documentation/cloud/data-rate-limits) which deny access to the API when you overload it with too many requests.

Perhaps the framework can provide some kind of mechanism where you can create groups (and possibly sub-groups) with objects (devices). These groups then have properties like time based limits and allow for operations (jobs) that take into account the defined group limits.

Reply:
The details of such a generic throttling mechanism can be a little tricky...

There are easy ones, e.g. the minimal delay which could be made configurable.

Then there is to distinguish whether the throttling should happen per bridge or per thing - still doable if it is completely agnostic to what it actually does, this would have to be handed over by `Runnables`. 

Then you will have to distinguish whether whatever-the-throttles-is-instructed-to-do should _really_ happen delayed (if a delay is required at all) or rather some of the executions should be skipped. While delaying might make sense in some cases (e.g. if I switch on a group of lights, I want every light to be on in the end, even if that's not possible at the same time) while sometimes it is rather not wanted (e.g. if I toggle a lamp a couple of times quickly, I might not want it to blink for the next minute but rather as soon as possible reach the final state). Unless I want to morse to my neighbors, then of course the blinking is intended 😉 

Just skipping the congested executions also might not be a suitable solution, because maybe only some command equal out each other. Others might be important. 

This all might still be doable with configurable implementations of different optimization strategies and callbacks for doing the "equalling out" calculations. 

But all that seems a little overdone compared to the few lines of code that are similar in LIFX and Trådfri. I would recommend waiting a little, collecting other use-cases before suggesting to implement a generic solution. Especially since I really consider this a bug in the Trådfri gateway that ideally should be fixed there - don't get me wrong, I'm kind of okay with having the workaround here, but that's what it is. I would not want to establish this as "best practice" or anything.